### PR TITLE
Add app unit tests using vitest

### PR DESCRIPTION
**Pull request by rijkvanzanten**
`vitest` seems ready (enough) for production use, so it's (more than overdue) time to start adding tests to the app. 

This PR sets up the basic vitest boilerplates, and adds as many unit tests as I can in the time I have rn.

also adding TSDoc comments to every util function I'm testing, so we can start playing with TypeDoc to auto-generate developer docs.

---

There's a handful of _questionable_ changes here that we should double check:

- https://github.com/directus/directus/pull/14583/commits/f42c44da537f4ec5ae01a22539c7d32c0c5ecdfc
- https://github.com/directus/directus/pull/14583/commits/e478a3a7aa9beb813cd611f02e8848f51b42838b

### CODE REVIEW

**Comment by rijkvanzanten** [comment]
@directus/team / @paescuj I've personally gone back-and-forth on this, so I'd love to hear your thoughts:

Do you prefer to keep tests in a separate folder:

```
src/
  utils/
    example.ts

tests/
  utils/
    example.test.ts
```

or next to the original file?

```
src
  utils/
    example.ts
    example.test.ts
```

I'm torn between _"Tests aren't part of the "source code" of the app, so keeping them separate makes sense"_ and _"Tests next to the source make it super obvious which ones are missing, and act as a reminder to keep them up to date"_

**Comment by paescuj** [comment]
I was about to propose "vitest" for the Directus app 😄 Although I've no personal experience with it yet, I think it is a good fit 🚀

> @directus/team / @paescuj I've personally gone back-and-forth on this, so I'd love to hear your thoughts:
> 
> Do you prefer to keep tests in a separate folder:
> 
> ```
> src/
>   utils/
>     example.ts
> 
> tests/
>   utils/
>     example.test.ts
> ```
> 
> or next to the original file?
> 
> ```
> src
>   utils/
>     example.ts
>     example.test.ts
> ```
> 
> I'm torn between _"Tests aren't part of the "source code" of the app, so keeping them separate makes sense"_ and _"Tests next to the source make it super obvious which ones are missing, and act as a reminder to keep them up to date"_

This is one of those hard questions because there is no "right" or "wrong"...

Personally, I'd go with the second option mainly because I think it makes devs life easier. This is especially the case for larger code bases like the Directus app:

* Easier to locate the tests files -> As you wrote: _"Tests next to the source make it super obvious which ones are missing, and act as a reminder to keep them up to date"_
* Imagine you want to update the folder structure when using the first option -> This means you'd need to update it in three places: `src` folder & `tests` folder and `import` in the corresponding file.

I think the second option works well as long as there is exactly one test file for each implementation file (which is usually the case with unit tests, right?).
That being said, IMO it's not a big deal if we should find out that the chosen option doesn't work for us as as we can easily switch between them (should more or less be a `bash` one-liner).

**Comment by br41nslug** [comment]
> Do you prefer to keep tests in a separate folder or next to the original file?

I think the second one has my preference so we don't have to do these deeply relative imports `import { addQueryToPath } from '../../src/utils/add-query-to-path';`. Doing tests in the same folder would make it easier to keep the tests in sync and up-to-date with refactors.

File: app/tsconfig.json [hunk 1]
```diff
@@ -12,14 +12,14 @@
...
+		"types": ["vite/client"],
 		"paths": {
 			"@/*": ["src/*"],
 			"preact/*": ["stub/empty.d.ts"],
 			"tslib": ["node_modules/tslib/tslib.d.ts"]
 		},
 		"lib": ["esnext", "dom", "dom.iterable", "scripthost"]
 	},
-	"include": ["src/**/*.ts", "src/**/*.vue"],
+	"include": ["src/**/*.ts", "src/**/*.vue", "tests/**/*.ts"],
```
Comment (paescuj) [inline app/tsconfig.json:23]:
```suggestion
	"include": ["src/**/*.ts", "src/**/*.vue"],
```

Not required anymore

**Pull request merged**